executor/src/witgen/data_structures/identity.rs

```diff
@@ -0,0 +1,430 @@
+    /// the negation has been removed.
+    pub multiplicity: Option<AlgebraicExpression<T>>,
+}
+
+impl<T: FieldElement> BusSend<T> {
+    /// Tries to find the matching receive, which should succeed iff. the send has
+    /// a static interaction ID.
+    pub fn try_match_static<'a>(
+        &self,
+        receives: &'a BTreeMap<T, BusReceive<T>>,
+    ) -> Option<&'a BusReceive<T>> {
+        Some(&receives[&self.interaction_id()?])
+    }
+
+    /// Returns the interaction ID if it is a number.
+    /// Sends and receives can be matched by matching the interaction ID.
+    pub fn interaction_id(&self) -> Option<T> {
+        match &self.interaction_id {
+            AlgebraicExpression::Number(id) => Some(*id),
+            _ => None,
+        }
+    }
+}
+
+impl<T: FieldElement> BusReceive<T> {
+    /// Returns whether the receive has an unconstrained multiplicity.
+    /// If it does not, this typically means that the multiplicity is binary constraint.
+    /// Unconstrained receives are equivalent to the RHS of a lookup;
+    /// binary-constrained receives are equivalent to the RHS of a permutation.
+    pub fn is_unconstrained(&self) -> bool {
+        // TODO: This always works in practice, but we should properly check the
+        // range constraints on the multiplicity column.
+        self.multiplicity.as_ref() != Some(&self.selected_tuple.selector)
+    }
+}
+
+impl<T> Children<AlgebraicExpression<T>> for BusSend<T> {
+    fn children_mut(&mut self) -> Box<dyn Iterator<Item = &mut AlgebraicExpression<T>> + '_> {
+        self.selected_tuple.children_mut()
+    }
+    fn children(&self) -> Box<dyn Iterator<Item = &AlgebraicExpression<T>> + '_> {
+        self.selected_tuple.children()
+    }
+}
+
+impl<T> Children<AlgebraicExpression<T>> for BusReceive<T> {
+    fn children_mut(&mut self) -> Box<dyn Iterator<Item = &mut AlgebraicExpression<T>> + '_> {
+        Box::new(
+            self.selected_tuple
+                .children_mut()
+                .chain(self.multiplicity.iter_mut()),
+        )
+    }
+    fn children(&self) -> Box<dyn Iterator<Item = &AlgebraicExpression<T>> + '_> {
+        Box::new(
+            self.selected_tuple
+                .children()
+                .chain(self.multiplicity.iter()),
+        )
+    }
+}
+
+impl<T: fmt::Display + fmt::Debug> fmt::Display for BusSend<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "BusSend(tuple={}, interaction_id={})",
+            self.selected_tuple, self.interaction_id
+        )
+    }
+}
+
+impl<T: fmt::Display + fmt::Debug> fmt::Display for BusReceive<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let multiplicity = match &self.multiplicity {
+            Some(m) => m.to_string(),
+            None => "None".to_string(),
+        };
+        write!(
+            f,
+            "BusReceive(tuple={}, multiplicity={}, interaction_id={})",
```

Comment:
Should we use `BusReceive{tuple: {}, multiplicity: {}...}`?
Also what about just `Send` and `Receive`? (for display purposes)

Reply:
I renamed it to `Send` and `Receive`. The dots are because there is also an ID? I personally find them more confusing than informative...

Reply:
The dots are because I was lazy to type it out in full

Reply:
Added the identity ID.